### PR TITLE
feat: add endpoints health checker

### PR DESCRIPTION
**Pull request by rkolpakov**

### CODE REVIEW

File: src/web3py/extensions/health_checker.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+        if all(id == chain_ids[0] for id in chain_ids):
+            return chain_ids[0]
+
+        raise ValueError(f'Different chain ids detected: {results}')
+
+    def get_hosts(self):
+        raise NotImplementedError()
+
+    def _health_check_call(self, host: any):
```
Comment (F4ever) [inline src/web3py/extensions/health_checker.py:26]: add annotations

File: src/web3py/extensions/health_checker.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+                raise RequestsConnectionError(f"Provider doesn't respond: {endpoint}") from exc
+
+        chain_ids = [r[1] for r in results]
+
+        if all(id == chain_ids[0] for id in chain_ids):
+            return chain_ids[0]
+
+        raise ValueError(f'Different chain ids detected: {results}')
+
+    def get_hosts(self):
```
Comment (F4ever) [inline src/web3py/extensions/health_checker.py:23]: add annotations

File: src/providers/http_provider.py [hunk 3]
```diff
@@ -151,3 +152,6 @@ def _get_without_fallbacks(
             meta = {}
 
         return data, meta
+
+    def get_hosts(self):
+        return list(map(lambda x: [x, x], self.hosts))
```
Comment (F4ever) [inline src/providers/http_provider.py:157]: better
`return [[host, host] for host in self.hosts]`

File: src/web3py/extensions/fallback.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from web3_multi_provider import FallbackProvider
+from src.web3py.extensions.health_checker import HealthChecker
+from web3 import Web3
+
+
+class FallbackProviderModule(FallbackProvider, HealthChecker):
+
+    def get_hosts(self):
+        return list(map(lambda p: [p, p.endpoint_uri], self._providers))
```
Comment (F4ever) [inline src/web3py/extensions/fallback.py:9]: [[provider, provider.endpoint_uri] for provider in self._providers]

File: src/web3py/extensions/health_checker.py (hunk 1 above)
Comment (F4ever) [reply src/web3py/extensions/health_checker.py:26]: get _chain_id

Comment (F4ever) [reply src/web3py/extensions/health_checker.py:26]: And return here only chain_id without host

File: src/main.py [hunk 5]
```diff
@@ -101,17 +102,17 @@ def check():
...
-    if any(execution_chain_id != chain_id for chain_id in [*chain_ids, consensus_chain_id, keys_api_chain_id]):
+def providers_health_check(web3: Web3, cc: ConsensusClientModule, kac: KeysAPIClientModule):
+    keys_api_chain_id = kac.health_check()
+    consensus_chain_id = cc.health_check()
+
+    el_provider = cast(FallbackProviderModule, web3.provider)
+
+    execution_chain_id = el_provider.health_check()
+
+    if not execution_chain_id == consensus_chain_id == keys_api_chain_id:
```
Comment (F4ever) [inline src/main.py:113]: Maybe
`if execution_chain_id == consensus_chain_id == keys_api_chain_id: return`

File: src/modules/checks/suites/common.py [hunk 6]
```diff
@@ -22,11 +21,6 @@ def ejector(web3, skip_locator):
     return Ejector(web3)
 
 
-def check_providers_chain_ids(web3):
```
Comment (F4ever) [inline src/modules/checks/suites/common.py:25]: Return back check_providers_chain_ids

Comment (F4ever) [reply src/modules/checks/suites/common.py:25]: and use inside your method

File: src/main.py [hunk 7]
```diff
@@ -101,19 +102,18 @@ def check():
...
-        Web3.to_int(hexstr=provider.make_request("eth_chainId", []).get('result'))
-        for provider in cast(FallbackProvider, web3.provider)._providers  # type: ignore[attr-defined] # pylint: disable=protected-access
-    ]
-    keys_api_chain_id = web3.kac.get_status().chainId
-    if any(execution_chain_id != chain_id for chain_id in [*chain_ids, consensus_chain_id, keys_api_chain_id]):
-        raise ValueError('Different chain ids detected:\n'
-                         f'Execution chain ids: {", ".join(map(str, chain_ids))}\n'
-                         f'Consensus chain id: {consensus_chain_id}\n'
-                         f'Keys API chain id: {keys_api_chain_id}\n')
+def providers_health_check(web3: Web3, cc: ConsensusClientModule, kac: KeysAPIClientModule):
```
Comment (F4ever) [inline src/main.py:105]:
```suggestion
def check_providers_chain_ids(web3: Web3, cc: ConsensusClientModule, kac: KeysAPIClientModule):
```

File: src/web3py/extensions/fallback.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+from web3_multi_provider import FallbackProvider
+from src.web3py.extensions.consistency import ProviderConsistencyModule
+from web3 import Web3
+
+
+class FallbackProviderModule(FallbackProvider, ProviderConsistencyModule):
+
+    def get_all_hosts(self):
```
Comment (F4ever) [inline src/web3py/extensions/fallback.py:8]: annotations

File: src/web3py/extensions/consistency.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+from typing import Any
+from abc import abstractmethod, ABC
+from requests.exceptions import ConnectionError as RequestsConnectionError
+
+
+class ProviderConsistencyModule(ABC):
+    """
+    A class that provides HTTP providers with the ability to check if
+    provided hosts are alive and to compare chain IDs when multiple providers are used.
```
Comment (F4ever) [inline src/web3py/extensions/consistency.py:9]:
```suggestion
    provided hosts are alive and chain ids are same.
```

File: src/providers/http_provider.py [hunk 10]
```diff
@@ -29,7 +29,7 @@ def __init__(self, *args, status: int, text: str):
         super().__init__(*args)
 
 
-class HTTPProvider(ABC):
+class HTTPProvider(ProviderConsistencyModule):
```
Comment (F4ever) [inline src/providers/http_provider.py:32]: Make this class abstract

File: src/web3py/extensions/consistency.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        """
+        Returns a list of hosts and URIs to be health checked.
+
+        HTTP provider returns URI string.
+        Web3 provider returns Provider instance.
+        """
+        raise NotImplementedError("get_all_hosts should be implemented")
+
+    @abstractmethod
+    def _chain_id(self, host) -> int:
```
Comment (F4ever) [inline src/web3py/extensions/consistency.py:43]:
```suggestion
    def get_chain_id(self, host) -> int:
```

Comment (F4ever) [reply src/web3py/extensions/consistency.py:43]: or _get_chain_id

File: src/providers/http_provider.py [hunk 12]
```diff
@@ -151,3 +151,9 @@ def _get_without_fallbacks(
             meta = {}
 
         return data, meta
+
+    def get_all_hosts(self):
+        return list(map(lambda host: [host, host], self.hosts))
+
+    def _chain_id(self, host) -> int:
```
[the remark this answers is not in the document]
Comment (rkolpakov) [reply src/providers/http_provider.py]: we can't do that because this class is used in tests

**Review by skhomuti — APPROVED** [review]

**Review by vgorkavenko — APPROVED** [review]

**Pull request merged**